File: internal/game/scoresheet.go
package game

type ScoreTotal struct {
	Total int
}

type ScoreSheet struct {
	Yellow *YellowScoreArea
	Green  *GreenScoreArea
	Orange *OrangeScoreArea
	Purple *PurpleScoreArea
	Blue   *BlueScoreArea
	Bonus  *BonusArea
}

type YellowScoreArea struct {
	ScoreTotal
	Columns [6][]bool // 6 columns, each with numbers 1-6
}

type GreenScoreArea struct {
	ScoreTotal
	Numbers []bool // 11 numbers: 2,3,4,5,6,7,8,9,10,11,12
}

type OrangeScoreArea struct {
	ScoreTotal
	Numbers []int // 11 spaces for any numbers
}

type PurpleScoreArea struct {
	ScoreTotal
	Numbers []bool // 11 numbers: 1-11, with special 6 reset
}

type BlueScoreArea struct {
	ScoreTotal
	Numbers []bool // 11 numbers: 1-11
}

type BonusArea struct {
	ScoreTotal // Calculated as lowest_section_score * FoxCount
	FoxCount   int
}

func NewScoreSheet() *ScoreSheet {
	return &ScoreSheet{
		Yellow: &YellowScoreArea{
			Columns: [6][]bool{
				make([]bool, 6), // Column 1: numbers 1-6
				make([]bool, 6), // Column 2: numbers 1-6
				make([]bool, 6), // Column 3: numbers 1-6
				make([]bool, 6), // Column 4: numbers 1-6
				make([]bool, 6), // Column 5: numbers 1-6
				make([]bool, 6), // Column 6: numbers 1-6
			},
		},
		Green: &GreenScoreArea{
			Numbers: make([]bool, 11),
		},
		Orange: &OrangeScoreArea{
			Numbers: make([]int, 11),
		},
		Purple: &PurpleScoreArea{
			Numbers: make([]bool, 11),
		},
		Blue: &BlueScoreArea{
			Numbers: make([]bool, 11),
		},
		Bonus: &BonusArea{
			FoxCount: 0,
		},
	}
}

func (ss *ScoreSheet) GetTotalScore() int {
	return ss.Yellow.Total + ss.Green.Total + ss.Orange.Total + ss.Purple.Total + ss.Blue.Total + ss.Bonus.Total
}

func (ss *ScoreSheet) CalculateBonus() {
	// Calculate bonus as lowest section score * fox count
	sections := []int{ss.Yellow.Total, ss.Green.Total, ss.Orange.Total, ss.Purple.Total, ss.Blue.Total}

	// Find lowest non-zero section
	lowest := 0
	for _, section := range sections {
		if section > 0 && (lowest == 0 || section < lowest) {
			lowest = section
		}
	}

	ss.Bonus.Total = lowest * ss.Bonus.FoxCount
}

func calculateYellowScore(yellow *YellowScoreArea) int {
	score := 0
	for col := range 6 {
		columnComplete := true
		for row := range 6 {
			if !yellow.Columns[col][row] {
				columnComplete = false
				break
			}
		}
		if columnComplete {
			score += (col + 1) * (col + 1)
		}
	}
	return score
}

func calculateGreenScore(green *GreenScoreArea) int {
	score := 0
	consecutiveCount := 0

	for i := range green.Numbers {
		if green.Numbers[i] {
			consecutiveCount++
		} else {
			if consecutiveCount > 0 {
				score += consecutiveCount * consecutiveCount
			}
			consecutiveCount = 0
		}
	}

	if consecutiveCount > 0 {
		score += consecutiveCount * consecutiveCount
	}

	return score
}

func calculateOrangeScore(orange *OrangeScoreArea) int {
	score := 0
	for _, num := range orange.Numbers {
		if num > 0 {
			score += num
		}
	}
	return score
}

func calculatePurpleScore(purple *PurpleScoreArea) int {
	score := 0
	for i, marked := range purple.Numbers {
		if marked {
			if i < 5 { // Numbers 1-5
				score += i + 1
			} else if i == 5 { // Number 6
				score += 6
			} else { // Numbers 7-11
				score += i + 1
			}
		}
	}
	return score
}

func calculateBlueScore(blue *BlueScoreArea) int {
	score := 0
	markedCount := 0

	for _, marked := range blue.Numbers {
		if marked {
			markedCount++
		}
	}

	switch markedCount {
	case 1:
		score = 1
	case 2:
		score = 3
	case 3:
		score = 6
	case 4:
		score = 10
	case 5:
		score = 15
	case 6:
		score = 21
	case 7:
		score = 28
	case 8:
		score = 36
	case 9:
		score = 45
	case 10:
		score = 55
	case 11:
		score = 66
	}

	return score
}

func calculateBonusScore(bonus *BonusArea) int {
	return bonus.Total
}

File: lib/models/score_sheet.dart
class ScoreSheet {
  YellowScoreArea yellow;
  GreenScoreArea green;
  OrangeScoreArea orange;
  PurpleScoreArea purple;
  BlueScoreArea blue;
  BonusArea bonus;

  ScoreSheet()
      : yellow = YellowScoreArea(),
        green = GreenScoreArea(),
        orange = OrangeScoreArea(),
        purple = PurpleScoreArea(),
        blue = BlueScoreArea(),
        bonus = BonusArea();

  int get totalScore =>
      yellow.total +
      green.total +
      orange.total +
      purple.total +
      blue.total +
      bonus.total;

  void calculateBonus() {
    List<int> sections = [
      yellow.total,
      green.total,
      orange.total,
      purple.total,
      blue.total
    ];

    int lowest = 0;
    for (int section in sections) {
      if (section > 0 && (lowest == 0 || section < lowest)) {
        lowest = section;
      }
    }

    bonus.total = lowest * bonus.foxCount;
  }

  Map<String, dynamic> toMap() {
    return {
      'yellow_total': yellow.total,
      'green_total': green.total,
      'orange_total': orange.total,
      'purple_total': purple.total,
      'blue_total': blue.total,
      'fox_count': bonus.foxCount,
      'bonus': bonus.total,
    };
  }
}

class YellowScoreArea {
  int total;
  List<List<bool>> columns;

  YellowScoreArea()
      : total = 0,
        columns = List.generate(6, (_) => List.filled(6, false));

  int calculateScore() {
    int score = 0;
    for (int col = 0; col < 6; col++) {
      bool columnComplete = true;
      for (int row = 0; row < 6; row++) {
        if (!columns[col][row]) {
          columnComplete = false;
          break;
        }
      }
      if (columnComplete) {
        score += (col + 1) * (col + 1);
      }
    }
    total = score;
    return score;
  }
}

class GreenScoreArea {
  int total;
  List<bool> numbers;

  GreenScoreArea()
      : total = 0,
        numbers = List.filled(11, false);

  int calculateScore() {
    int score = 0;
    int consecutiveCount = 0;

    for (int i = 0; i < numbers.length; i++) {
      if (numbers[i]) {
        consecutiveCount++;
      } else {
        if (consecutiveCount > 0) {
          score += consecutiveCount * consecutiveCount;
        }
        consecutiveCount = 0;
      }
    }

    if (consecutiveCount > 0) {
      score += consecutiveCount * consecutiveCount;
    }

    total = score;
    return score;
  }
}

class OrangeScoreArea {
  int total;
  List<int> numbers;

  OrangeScoreArea()
      : total = 0,
        numbers = List.filled(11, 0);

  int calculateScore() {
    int score = 0;
    for (int num in numbers) {
      if (num > 0) {
        score += num;
      }
    }
    total = score;
    return score;
  }
}

class PurpleScoreArea {
  int total;
  List<bool> numbers;

  PurpleScoreArea()
      : total = 0,
        numbers = List.filled(11, false);

  int calculateScore() {
    int score = 0;
    for (int i = 0; i < numbers.length; i++) {
      if (numbers[i]) {
        if (i < 5) {
          score += i + 1;
        } else if (i == 5) {
          score += 6;
        } else {
          score += i + 1;
        }
      }
    }
    total = score;
    return score;
  }
}

class BlueScoreArea {
  int total;
  List<bool> numbers;

  BlueScoreArea()
      : total = 0,
        numbers = List.filled(11, false);

  int calculateScore() {
    int markedCount = 0;
    for (bool marked in numbers) {
      if (marked) markedCount++;
    }

    int score;
    switch (markedCount) {
      case 1:
        score = 1;
        break;
      case 2:
        score = 3;
        break;
      case 3:
        score = 6;
        break;
      case 4:
        score = 10;
        break;
      case 5:
        score = 15;
        break;
      case 6:
        score = 21;
        break;
      case 7:
        score = 28;
        break;
      case 8:
        score = 36;
        break;
      case 9:
        score = 45;
        break;
      case 10:
        score = 55;
        break;
      case 11:
        score = 66;
        break;
      default:
        score = 0;
    }

    total = score;
    return score;
  }
}

class BonusArea {
  int total;
  int foxCount;

  BonusArea()
      : total = 0,
        foxCount = 0;
}
